File: examples/java/patterns/saga-pattern/pitfall2-in-memory-state.java
// WRONG - timestamp outside a step, changes on replay
long timestamp = System.currentTimeMillis();
compensations.add(new Compensation("cancel",
    () -> cancelWithTimestamp((String) reservation.get("id"), timestamp)
));

// CORRECT - data from step return value, stable on replay
Map<String, Object> reservation = context.step("reserve", Map.class,
    ctx -> reserveInventory(orderId));
compensations.add(new Compensation("cancel-reservation",
    () -> cancelReservation((String) reservation.get("id"))
));

File: examples/python/patterns/saga-pattern/pitfall2-in-memory-state.py
# WRONG - timestamp outside a step, changes on replay
import time
timestamp = time.time()
compensations.append(('cancel-reservation', lambda: cancel_reservation(timestamp)))  # timestamp changes on replay!

# CORRECT - data from step return value, stable on replay
reservation = context.step(reserve_inventory(order_id))
compensations.append(('cancel-reservation', lambda: cancel_reservation(reservation['id'])))